File: app/src/main/java/com/joeshuff/dddungeongenerator/generator/monsters/Bestiary.java
package com.joeshuff.dddungeongenerator.generator.monsters;

import android.content.Context;
import com.google.gson.Gson;
import com.joeshuff.dddungeongenerator.generator.dungeon.Modifier;
import org.json.JSONArray;
import org.json.JSONObject;

import java.io.IOException;
import java.io.InputStream;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Random;
import java.util.stream.Collectors;

/**
 * Creates various Monster objects filling in params from the Monster class
 * Creates lists for each 'theme' and adds the monsters to their respective themes
 *
 */

public class Bestiary {

	private static List<Monster> monsterList = new ArrayList<>();

	List<String> bossSize = Arrays.asList("large", "gargantuan", "huge");

	private static String loadJSONFromAsset(Context c) {
		String json = null;
		try {
			InputStream is = c.getAssets().open("monsters.json");
			int size = is.available();
			byte[] buffer = new byte[size];
			is.read(buffer);
			is.close();
			json = new String(buffer, "UTF-8");
		} catch (IOException ex) {
			ex.printStackTrace();
			return null;
		}
		return json;
	}

	public static void launchBestiary(Context c) {
		try {
			JSONArray jsonObject = new JSONArray(loadJSONFromAsset(c));

			for (int i = 0; i < jsonObject.length(); i ++) {
				monsterList.add(new Gson().fromJson(jsonObject.getJSONObject(i).toString(), Monster.class));
			}
		} catch (Exception e) {
			e.printStackTrace();
		}
	}

	public static List<String> getTypes() {
		List<String> res = new ArrayList<>();

		for (Monster m : monsterList) {
			if (res.contains(m.getType())) continue;
			res.add(m.getType());
		}

		return res;
	}

	public static Monster getBoss(Random rnd, Modifier modifier, boolean ignorePreferences) {
		//A boss is defined as a monster of size
		List<String> preferredTypes = modifier.getPreferredMonsters().stream().map(monster_class -> monster_class.getType()).collect(Collectors.toList());

		if (preferredTypes.isEmpty() || ignorePreferences) {
			preferredTypes.clear();
			preferredTypes.addAll(Arrays.stream(MonsterClass.MONSTER_CLASS.values()).map(monster_class -> monster_class.getType()).collect(Collectors.toList()));
		}

		List<String> ignoreTypes = modifier.getBlockedMonsters().stream().map(monster_class -> monster_class.getType()).collect(Collectors.toList());

		List<Monster> gargantuanPreferred = monsterList.stream()
				.filter(monster -> monster.getSize().equalsIgnoreCase("gargantuan") &&
						preferredTypes.contains(monster.getType()) &&
						!ignoreTypes.contains(monster.getType()))
				.collect(Collectors.toList());

		List<Monster> hugePreferred = monsterList.stream()
				.filter(monster -> monster.getSize().equalsIgnoreCase("huge") &&
						preferredTypes.contains(monster.getType()) &&
						!ignoreTypes.contains(monster.getType()))
				.collect(Collectors.toList());

		List<Monster> largePreferred = monsterList.stream()
				.filter(monster -> monster.getSize().equalsIgnoreCase("large") &&
						preferredTypes.contains(monster.getType()) &&
						!ignoreTypes.contains(monster.getType()))
				.collect(Collectors.toList());

		List<Monster> allPreferred = new ArrayList<>();

		allPreferred.addAll(gargantuanPreferred);
		allPreferred.addAll(hugePreferred);
		allPreferred.addAll(largePreferred);

		if (allPreferred.size() == 0) return getBoss(rnd, modifier, true);

		return allPreferred.get(rnd.nextInt(allPreferred.size()));
	}

	public static Monster getMonster(Random rnd, Modifier modifier, boolean ignorePreferred) {
		//A boss is defined as a monster of size
		List<String> preferredTypes = modifier.getPreferredMonsters().stream().map(monster_class -> monster_class.getType()).collect(Collectors.toList());

		if (preferredTypes.isEmpty() || ignorePreferred) {
			preferredTypes.clear();
			preferredTypes.addAll(Arrays.stream(MonsterClass.MONSTER_CLASS.values()).map(monster_class -> monster_class.getType()).collect(Collectors.toList()));
		}

		List<String> ignoreTypes = modifier.getBlockedMonsters().stream().map(monster_class -> monster_class.getType()).collect(Collectors.toList());

		List<Monster> tinyPreferred = monsterList.stream()
				.filter(monster -> monster.getSize().equalsIgnoreCase("tiny") &&
						preferredTypes.contains(monster.getType()) &&
						!ignoreTypes.contains(monster.getType()))
				.collect(Collectors.toList());

		List<Monster> smallPreferred = monsterList.stream()
				.filter(monster -> monster.getSize().equalsIgnoreCase("small") &&
						preferredTypes.contains(monster.getType()) &&
						!ignoreTypes.contains(monster.getType()))
				.collect(Collectors.toList());

		List<Monster> mediumPreferred = monsterList.stream()
				.filter(monster -> monster.getSize().equalsIgnoreCase("medium") &&
						preferredTypes.contains(monster.getType()) &&
						!ignoreTypes.contains(monster.getType()))
				.collect(Collectors.toList());

		List<Monster> allPreferred = new ArrayList<>();

		allPreferred.addAll(tinyPreferred);
		allPreferred.addAll(smallPreferred);
		allPreferred.addAll(mediumPreferred);

		if (allPreferred.size() == 0) return getMonster(rnd, modifier, true);

		return allPreferred.get(rnd.nextInt(allPreferred.size()));
	}

}

File: app/src/main/java/com/joeshuff/dddungeongenerator/generator/monsters/Bestiary.kt
package com.joeshuff.dddungeongenerator.generator.monsters

import android.content.Context
import com.google.gson.Gson
import com.joeshuff.dddungeongenerator.generator.dungeon.Modifier
import com.joeshuff.dddungeongenerator.generator.monsters.MonsterClass.MONSTER_CLASS
import org.json.JSONArray
import java.io.IOException
import java.util.*
import kotlin.collections.ArrayList

object Bestiary {
    private val monsterList: ArrayList<Monster> = ArrayList()

    var bossSize = listOf("large", "gargantuan", "huge")

    private fun loadJSONFromAsset(c: Context): String? {
        return try {
            val inputstream = c.assets.open("monsters.json")
            val size = inputstream.available()
            val buffer = ByteArray(size)
            inputstream.read(buffer)
            inputstream.close()
            String(buffer, Charsets.UTF_8)
        } catch (ex: IOException) {
            ex.printStackTrace()
            null
        }
    }

    fun launchBestiary(c: Context) {
        try {
            val jsonObject = JSONArray(loadJSONFromAsset(c)?: "[]")
            for (i in 0 until jsonObject.length()) {
                monsterList.add(Gson().fromJson(jsonObject.getJSONObject(i).toString(), Monster::class.java))
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    fun getTypes(): List<String?>? {
        val res: MutableList<String?> = ArrayList()
        for (m in monsterList) {
            if (res.contains(m.type)) continue
            res.add(m.type)
        }
        return res
    }

    fun getMonster(rnd: Random, mod: Modifier, ignorePrefs: Boolean, boss: Boolean = false): Monster {
        //A boss is defined as a monster of size
        val preferredTypes: ArrayList<String> = ArrayList(mod.getPreferredMonsters().map { it.type })

        if (preferredTypes.isEmpty() || ignorePrefs) {
            preferredTypes.clear()
            preferredTypes.addAll(MONSTER_CLASS.values().map { it.type })
        }

        val ignoreTypes = mod.getBlockedMonsters().map { it.type }

        val preferred = monsterList.filter {
            ((boss && bossSize.contains(it.size?.toLowerCase())) || (!boss && !bossSize.contains(it.size?.toLowerCase()))) &&
                    preferredTypes.contains(it.type) &&
                    !ignoreTypes.contains(it.type)
        }

        return if (preferred.isEmpty()) getMonster(rnd, mod, true, boss) else preferred.random()
    }

    fun getBoss(rnd: Random, mod: Modifier, ignorePrefs: Boolean): Monster {
        return getMonster(rnd, mod, ignorePrefs, true)
    }
}